provider/openstack/provider.go

```diff
@@ -1459,20 +1467,37 @@ func (e *Environ) networksForInstance(args environs.StartInstanceParams) ([]nova
 	// Set the subnetID on the network for all networks.
 	// For each of the subnetIDs selected, create a port for each one.
 	subnetNetworks := make([]nova.ServerNetworks, 0, len(subnetIDForZone))
-	for _, subnetID := range subnetIDForZone {
-		var portID string
-		portID, err = e.networking.CreatePort(e.uuid, networkID, subnetID)
+	netInfo := make([]corenetwork.InterfaceInfo, len(subnetIDsForZone))
+	for i, subnetID := range subnetIDForZone {
+		var port *neutron.PortV2
+		port, err = e.networking.CreatePort(e.uuid, networkID, subnetID)
 		if err != nil {
 			break
 		}
 
-		logger.Infof("created new port %q connected to Openstack subnet %q", portID, subnetID)
+		logger.Infof("created new port %q connected to Openstack subnet %q", port.Id, subnetID)
 		subnetNetworks = append(subnetNetworks, nova.ServerNetworks{
 			NetworkId: networkID,
-			PortId:    portID,
+			PortId:    port.Id,
 		})
+
+		// We expect a single address,
+		// but for correctness we add all from the created port.
+		ips := make([]string, len(port.FixedIPs))
+		for j, fixedIP := range port.FixedIPs {
+			ips[j] = fixedIP.IPAddress
+		}
+
+		netInfo[i] = corenetwork.InterfaceInfo{
+			InterfaceName: fmt.Sprintf("eth%d", i),
+			MACAddress:    port.MACAddress,
+			Addresses:     corenetwork.NewProviderAddresses(ips...),
+			ConfigType:    corenetwork.ConfigDHCP,
+		}
 	}
 
+	err = cloudCfg.AddNetworkConfig(netInfo)
```

Comment:
Excellent, it's a shame that we don't invert this, so adding a network config isn't magic.